Server version param replication fix

marcinmdev opened the pull request:
|      Q           |   A
|------------- | -----------
| Type            | bug
| BC Break     | no
| Fixed issues | -

#### Summary
Sets serverVersion from database url to params when using replication scheme. Prevents unnecessary connections to database with every request, even when getting data from cache.

mfrankiewicz commented:
very important fix, this non-obvious behavior caused a lot of debugging and investigations during load tests

derrabus commented:
Can we add a test for this change?

marcinmdev commented:
Added test.

derrabus commented:
Please run `./vendor/bin/phpcbf` and commit the changes.

marcinmdev commented:
Done.

greg0ire reviewed (dismissed):
Needs to be squash-merged

greg0ire commented:
Little concern: can a primary and its replica have different server versions? And what would happen in that case?

morozov commented:
> can a primary and its replica have different server versions?

They can as long as the replication protocol is compatible in the direction of replication ([ref](https://dev.mysql.com/doc/mysql-replication-excerpt/5.7/en/replication-compatibility.html)). E.g. one can replicate MySQL 5 to MySQL 8 which from the DBAL perspective may be two different platforms. I believe the same will work for replication between MySQL and MariaDB.

greg0ire commented:
@marcinmwiosna my understanding of this PR is that it "extracts" the server version from the primary config to generalize it at the upper level. Then that server version is used to connect to any replica. Isn't that a wrong generalization, given what Sergei mentioned? What happens if 2 different server versions are specified for both the primary and its replica?

marcinmdev commented:
@greg0ire It only extracts server version from DB URL/DSN. It has nothing to do with further processing, which is done elsewhere.
> What happens if 2 different server versions are specified for both the primary and its replica?

I think it is separate issue from the one I mentioned in this PR.

greg0ire commented:
> I think it is separate issue from the one I mentioned in this PR.

It won't be if the behavior is not the same before and after your PR.

marcinmdev commented:
I'm using doctrine with Symfony.
There are couple issues there:
* Symfony does not allow explicitly setting server version for replica in their config, only for primary
* on the main DBAL branch (meaning when excluding my PR) if we use URL with different versions then the primary will be ignored and taken from database (which means database query with every single request)

I don't really know what is the intended behavior in there. There is clear discrepancy between explicitly setting server version through config and using url parsing.

greg0ire commented:
There are a bit too many things we don't know here IMO:
- We don't know what happens when the server version differs from primary to replica.
- We don't know why setting the server version in the URL is not sufficient for the DBAL not to ask it every time it tries to connect, and so we don't know if that's the thing that should be fixed instead.
- We don't know what the intended behavior is.

morozov commented on tests/Doctrine/Tests/DBAL/DriverManagerTest.php:
```diff
@@ -144,7 +145,7 @@ public function testDatabaseUrlPrimaryReplica(): void
         $options = [
             'driver' => 'pdo_mysql',
-            'primary' => ['url' => 'mysql://foo:bar@localhost:11211/baz'],
+            'primary' => ['url' => 'mysql://foo:bar@localhost:11211/baz?serverVersion=11.2'],
```
Please do not modify an existing test. Add a new one instead.

morozov commented on tests/Doctrine/Tests/DBAL/DriverManagerTest.php:
```diff
@@ -156,30 +157,34 @@ public function testDatabaseUrlPrimaryReplica(): void
+            'serverVersion' => '11.2',
         ];
 
+        $expectedReplica = array_diff_key($expectedPrimary, ['serverVersion' => null]);
```
Please [refrain from putting logic in tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html).

marcinmdev commented:
@morozov I've rewritten the test.

marcinmdev commented:
> * We don't know why setting the server version in the URL is not sufficient for the DBAL not to ask it every time it tries to connect, and so we don't know if that's the thing that should be fixed instead.

@greg0ire This param is ignored in the code and this PR attempts to fix it.

> * We don't know what the intended behavior is.

I can't answer, as I'm not the original author here and I haven't found sufficient documentation. 

> We don't know what happens when the server version differs from primary to replica.

I assume that allowing different versions between replicas/primary could lead to unwanted/unpredictable behavior.

morozov commented:
> I've rewritten the test.

This test passes without the fix, so it probably doesn't cover your use case. Note that the `2.13.x` accepts only improvements to the upgrade path to DBAL 3. Bugfixes are accepted in `3.2.x`.

marcinmdev commented:
I've fixed the test.
To make the issue more clear:
https://github.com/doctrine/dbal/blob/d1e581da590d611c8699acff9848056b2403c05b/src/Connection.php#L382-L388

This code gets serverVersion from the root of $params. Using replication without fix from this PR means that serverVersion lies in "primary" array, and Connection does not see it as set and connects to DB every time.

morozov commented on tests/DriverManagerTest.php:
```diff
@@ -149,6 +149,51 @@ public function testDatabaseUrlPrimaryReplica(): void
+                ],
+                'serverVersion' => '11.2',
+            ],
+            array_intersect_key($params, ['primary' => null, 'replica' => null, 'serverVersion' => null])
```
Please [refrain from putting logic into tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html). Instead, provide the expected values statically.

morozov commented on tests/DriverManagerTest.php:
```diff
@@ -149,6 +149,51 @@ public function testDatabaseUrlPrimaryReplica(): void
+        self::assertInstanceOf(PDO\MySQL\Driver::class, $conn->getDriver());
+
+        $expected = [
+            'user'          => 'foo',
```
This doesn't seem related to the test scenario. Please make only the assertions that are relevant to the test.

morozov commented on tests/DriverManagerTest.php:
```diff
@@ -149,6 +149,51 @@ public function testDatabaseUrlPrimaryReplica(): void
+            'driver' => 'pdo_mysql',
+            'primary' => ['url' => 'mysql://foo:bar@localhost:11211/baz?serverVersion=11.2'],
+            'replica' => [
+                'replica1' => ['url' => 'mysql://foo:bar@localhost:11211/baz_replica?serverVersion=11.2'],
```
If the server version is specified both in the primary and replica URLs, which of them is expected to be passed to the replica driver?

marcinmdev replied:
Database platform seem to be shared between primary and replica. I assume that implementing ability to set different versions between them would require connection logic to be rewritten.

marcinmdev committed:
[HEAD] [set-server-version-from-url-when-using-replication] simplified test
Sets serverVersion from database url to params when using replication
scheme. Prevents unnecessary connections to database with every
request, even when getting data from cache.

morozov commented:
@marcinmwiosna your PR needs to be rebased.

marcinmdev commented:
I've squashed and cleaned up the commits

marcinmdev commented:
I've fixed test condition and docblock annotations to fulfill Psalm requirements.

morozov commented on tests/DriverManagerTest.php:
```diff
@@ -149,6 +149,28 @@ public function testDatabaseUrlPrimaryReplica(): void
+        $params = $conn->getParams();
+        self::assertInstanceOf(PDO\MySQL\Driver::class, $conn->getDriver());
+
+        self::assertArrayHasKey('serverVersion', $params);
```
What is the reason for this requirement? As I see it, without the fix, `$params` contains the 'primary' and 'replica.replica1' keys which contain the corresponding connection parameters including the server version:
```php
array(
  'driver' => 'pdo_mysql',
  'primary' => array(
    // ...
    'serverVersion' => '11.2',
  ),
  'replica' => array(
    'replica1' => array(
      // ...
      'serverVersion' => '11.2',
    ),
  ),
)
```

Why does it have to be also contained at the top level? What does top-level server version mean?

marcinmdev replied:
https://github.com/doctrine/dbal/pull/4900#issuecomment-1015189401
Practically, without fix in this PR, Doctrine connects to DB to verify server version with every run of the code (while using replication).

morozov replied:
This is clear. The question is, why doesn't Doctrine use the server version specified in the corresponding connection parameters?

morozov replied:
The problem seems to be here: https://github.com/doctrine/dbal/blob/e839cecc812fd04ec6c39fc865302b6a322d967d/src/Connection.php#L398-L401

The connection doesn't expect the structure of its parameters changed by the extending class and expects the server version at the top level.

Since the `PrimaryReplicaConnection` changes the structure of parameters, it should change the way of detecting the platform accordingly. The closest relevant non-private method is:

https://github.com/doctrine/dbal/blob/e839cecc812fd04ec6c39fc865302b6a322d967d/src/Connection.php#L282

marcinmdev replied:
I preferred changing the logic in DriverManager because it seemed to be the simplest way of fixing this issue.
Connection is not Replication aware and it would require more changes to its logic, which I deemed undesirable.

morozov replied:
I’ll defer further review to @beberlei since he owns this part of the codebase.

andyexeter commented:
Is there anything I can do help this PR progress? It'd be great to see this issue resolved.
